File: src/use-cases/auth/LoginUser.js
/**
 * LoginUser Use Case
 * Handles user authentication and JWT token generation
 */

const logger = require('../../../logger');

class LoginUser {
    /**
     * @param {IUserRepository} userRepository - User repository
     * @param {JwtService} jwtService - JWT service
     */
    constructor(userRepository, jwtService) {
        this.userRepository = userRepository;
        this.jwtService = jwtService;
    }

    /**
     * Execute the use case
     * @param {Object} request - Login request
     * @param {string} request.username - Username
     * @param {string} request.password - Password
     * @returns {Promise<Object>} Login result with user and token
     */
    async execute({ username, password }) {
        try {
            // Validate input
            if (!username || typeof username !== 'string') {
                throw new Error('Username is required');
            }

            if (!password || typeof password !== 'string') {
                throw new Error('Password is required');
            }

            const trimmedUsername = username.trim();

            // Find user by username
            const user = await this.userRepository.findByUsername(trimmedUsername);

            if (!user) {
                logger.warn('Login failed: user not found', { username: trimmedUsername });
                throw new Error('Invalid username or password');
            }

            // Verify password
            const isPasswordValid = await user.verifyPassword(password);

            if (!isPasswordValid) {
                logger.warn('Login failed: invalid password', {
                    userId: user.id,
                    username: user.username
                });
                throw new Error('Invalid username or password');
            }

            // Update last login timestamp
            await this.userRepository.updateLastLogin(user.id);

            // Generate JWT token (include isAdmin for frontend)
            const token = this.jwtService.sign({
                userId: user.id,
                username: user.username,
                isAdmin: user.isAdminUser()
            });

            logger.info('User logged in successfully', {
                userId: user.id,
                username: user.username,
                isAdmin: user.isAdminUser()
            });

            return {
                success: true,
                user: user.toPublicObject(),
                token
            };
        } catch (error) {
            logger.error('Login error', {
                username,
                error: error.message
            });

            throw error;
        }
    }
}

module.exports = LoginUser;

File: zapzap-rust/src/application/auth/login_user.rs
use std::sync::Arc;

use crate::domain::entities::User;
use crate::domain::repositories::UserRepository;
use crate::infrastructure::auth::{JwtService, PasswordService};

/// Login user input
pub struct LoginUserInput {
    pub username: String,
    pub password: String,
}

/// Login user output
pub struct LoginUserOutput {
    pub user: User,
    pub token: String,
}

/// Login user use case
pub struct LoginUser {
    user_repo: Arc<dyn UserRepository>,
    jwt_service: Arc<JwtService>,
}

impl LoginUser {
    pub fn new(user_repo: Arc<dyn UserRepository>, jwt_service: Arc<JwtService>) -> Self {
        Self {
            user_repo,
            jwt_service,
        }
    }

    pub async fn execute(&self, input: LoginUserInput) -> Result<LoginUserOutput, LoginError> {
        // Validate input
        if input.username.trim().is_empty() {
            return Err(LoginError::Validation("Username is required".into()));
        }
        if input.password.is_empty() {
            return Err(LoginError::Validation("Password is required".into()));
        }

        // Find user
        let user = self
            .user_repo
            .find_by_username(&input.username)
            .await?
            .ok_or(LoginError::InvalidCredentials)?;

        // Check if user is a bot
        if user.is_bot() {
            return Err(LoginError::InvalidCredentials);
        }

        // Verify password
        let password_hash = user
            .password_hash
            .as_ref()
            .ok_or(LoginError::InvalidCredentials)?;

        // A stored hash that cannot be read refuses the login, as Node's `bcrypt.compare`
        // answers false for it, rather than failing the request
        let valid = PasswordService::verify(&input.password, password_hash).unwrap_or_else(|e| {
            tracing::warn!("Unreadable password hash for user {}: {}", user.id, e);
            false
        });

        if !valid {
            return Err(LoginError::InvalidCredentials);
        }

        // Update last login. The hash is never rewritten on login: a bcrypt hash stays
        // bcrypt, an Argon2 one stays Argon2.
        self.user_repo.update_last_login(&user.id).await?;

        // Generate token
        let token = self
            .jwt_service
            .sign(&user.id, &user.username, user.is_admin)
            .map_err(|e| LoginError::Internal(e.to_string()))?;

        Ok(LoginUserOutput { user, token })
    }
}

/// Login error types
#[derive(Debug, thiserror::Error)]
pub enum LoginError {
    #[error("Validation error: {0}")]
    Validation(String),
    #[error("Invalid username or password")]
    InvalidCredentials,
    #[error("Internal error: {0}")]
    Internal(String),
    #[error("Repository error: {0}")]
    Repository(#[from] crate::domain::repositories::RepositoryError),
}
